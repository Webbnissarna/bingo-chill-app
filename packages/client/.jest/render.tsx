import React from "react";
import { render as rtlRender } from "@testing-library/react";

export function render(ui: React.ReactElement) {
  return rtlRender(<>{ui}</>);
}
