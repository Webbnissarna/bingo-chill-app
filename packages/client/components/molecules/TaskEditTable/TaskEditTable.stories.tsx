import type { StoryDefault } from "@ladle/react";
import TaskEditTable from ".";

export function Default() {
  return (
    <div className="bg-polarNight-1 p-5">
      <TaskEditTable tasks={[]} onTasksChanged={() => undefined} />
    </div>
  );
}

export default {
  title: "Molecules / Task Edit Table",
} satisfies StoryDefault;
