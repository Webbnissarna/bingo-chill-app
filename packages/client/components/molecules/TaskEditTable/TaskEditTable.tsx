/* import { MenuOutlined } from "@ant-design/icons";
import type { DragEndEvent } from "@dnd-kit/core";
import { DndContext } from "@dnd-kit/core";
import { restrictToVerticalAxis } from "@dnd-kit/modifiers";
import {
  arrayMove,
  SortableContext,
  useSortable,
  verticalListSortingStrategy,
} from "@dnd-kit/sortable";
import { CSS } from "@dnd-kit/utilities";

import { Table } from "antd";
import type { ColumnsType } from "antd/es/table";
import React, { useState } from "react";

interface DataType {
  key: string;
  name: string;
  age: number;
  address: string;
}

const columns: ColumnsType<DataType> = [
  {
    key: "sort",
  },
  {
    title: "Name",
    dataIndex: "name",
  },
  {
    title: "Age",
    dataIndex: "age",
  },
  {
    title: "Address",
    dataIndex: "address",
  },
];

interface RowProps extends React.HTMLAttributes<HTMLTableRowElement> {
  "data-row-key": string;
}

function Row({ children, ...props }: RowProps) {
  const {
    attributes,
    listeners,
    setNodeRef,
    setActivatorNodeRef,
    transform,
    transition,
    isDragging,
  } = useSortable({
    id: props["data-row-key"],
  });

  const style: React.CSSProperties = {
    ...props.style,
    transform: CSS.Transform.toString(transform && { ...transform, scaleY: 1 }),
    transition,
    ...(isDragging ? { position: "relative", zIndex: 9999 } : {}),
  };

  return (
    <tr {...props} ref={setNodeRef} style={style} {...attributes}>
      {React.Children.map(children, (child) => {
        if ((child as React.ReactElement).key === "sort") {
          return React.cloneElement(child as React.ReactElement, {
            children: (
              <MenuOutlined
                ref={setActivatorNodeRef}
                style={{ touchAction: "none", cursor: "move" }}
                {...listeners}
              />
            ),
          });
        }
        return child;
      })}
    </tr>
  );
}

export default function TaskEditTable(): JSX.Element {
  const [dataSource, setDataSource] = useState([
    {
      key: "1",
      name: "John Brown",
      age: 32,
      address:
        "Long text Long text Long text Long text Long text Long text Long text Long text Long text Long text Long text Long text Long text Long text Long text Long text Long text Long text Long text Long text Long text Long text Long text Long text Long text Long text Long text Long text Long text Long text Long text Long text",
    },
    {
      key: "2",
      name: "Jim Green",
      age: 42,
      address: "London No. 1 Lake Park",
    },
    {
      key: "3",
      name: "Joe Black",
      age: 32,
      address: "Sidney No. 1 Lake Park",
    },
  ]);

  const onDragEnd = ({ active, over }: DragEndEvent) => {
    if (active.id !== over?.id) {
      setDataSource((previous) => {
        const activeIndex = previous.findIndex((i) => i.key === active.id);
        const overIndex = previous.findIndex((i) => i.key === over?.id);
        return arrayMove(previous, activeIndex, overIndex);
      });
    }
  };

  return (
    <DndContext modifiers={[restrictToVerticalAxis]} onDragEnd={onDragEnd}>
      <SortableContext
        // rowKey array
        items={dataSource.map((i) => i.key)}
        strategy={verticalListSortingStrategy}
      >
        <Table
          components={{
            body: {
              row: Row,
            },
          }}
          rowKey="key"
          columns={columns}
          dataSource={dataSource}
        />
      </SortableContext>
    </DndContext>
  );
} */

import { MenuOutlined } from "@ant-design/icons";
import type { DragEndEvent } from "@dnd-kit/core";
import { DndContext } from "@dnd-kit/core";
import { restrictToVerticalAxis } from "@dnd-kit/modifiers";
import {
  arrayMove,
  SortableContext,
  useSortable,
  verticalListSortingStrategy,
} from "@dnd-kit/sortable";
import { CSS } from "@dnd-kit/utilities";

import { Table } from "antd";
import type { ColumnsType } from "antd/es/table";
import React, { useState } from "react";
import type { Task } from "@webbnissarna/bingo-chill-common/src/game/types";
import { TagsInput, TextButton, TextInputField } from "@/components/atoms";

interface DataType extends Task {
  key: string;
}

const columns: ColumnsType<DataType> = [
  {
    key: "sort",
  },
  {
    title: "Icon",
    dataIndex: "icon",
    render: () => <div className="w-12 h-12 bg-polarNight-0 rounded-md" />,
  },
  {
    title: "Name",
    dataIndex: "name",
    render: (v) => <TextInputField value={v} onChange={() => undefined} />,
  },
  {
    title: "Tags",
    dataIndex: "tags",
    render: (v) => (
      <TagsInput
        values={v}
        onChange={() => undefined}
        options={["A", "B", "C", "D"].map((x) => ({
          label: x,
          value: x.toLowerCase(),
        }))}
        placeholder="tags"
      />
    ),
  },
  {
    render: () => <TextButton text="Delete" onClick={() => undefined} />,
  },
];

interface RowProps extends React.HTMLAttributes<HTMLTableRowElement> {
  "data-row-key": string;
}

function Row({ children, ...props }: RowProps) {
  const {
    attributes,
    listeners,
    setNodeRef,
    setActivatorNodeRef,
    transform,
    transition,
    isDragging,
  } = useSortable({
    id: props["data-row-key"],
  });

  const style: React.CSSProperties = {
    ...props.style,
    transform: CSS.Transform.toString(transform && { ...transform, scaleY: 1 }),
    transition,
    ...(isDragging ? { position: "relative", zIndex: 9999 } : {}),
  };

  return (
    <tr {...props} ref={setNodeRef} style={style} {...attributes}>
      {React.Children.map(children, (child) => {
        if ((child as React.ReactElement).key === "sort") {
          return React.cloneElement(child as React.ReactElement, {
            children: (
              <MenuOutlined
                ref={setActivatorNodeRef}
                style={{ touchAction: "none", cursor: "move" }}
                {...listeners}
              />
            ),
          });
        }
        return child;
      })}
    </tr>
  );
}

export default function TaskEditTable(): JSX.Element {
  const [count, setCount] = useState(2);
  const [dataSource, setDataSource] = useState<DataType[]>([
    { key: "1", name: "Hello World", icon: "", tags: ["A", "B", "C"] },
    { key: "2", name: "Hello World 2", icon: "", tags: ["A", "B", "C"] },
  ]);

  const onDragEnd = ({ active, over }: DragEndEvent) => {
    if (active.id !== over?.id) {
      setDataSource((previous) => {
        const activeIndex = previous.findIndex((i) => i.key === active.id);
        const overIndex = previous.findIndex((i) => i.key === over?.id);
        return arrayMove(previous, activeIndex, overIndex);
      });
    }
  };

  const add = () => {
    setDataSource([
      ...dataSource,
      { key: `${count}`, name: "", icon: "", tags: [] },
    ]);
    setCount(count + 1);
  };

  return (
    <div className="flex flex-col gap-2">
      <DndContext modifiers={[restrictToVerticalAxis]} onDragEnd={onDragEnd}>
        <SortableContext
          items={dataSource.map((i) => i.key)}
          strategy={verticalListSortingStrategy}
        >
          <Table
            pagination={false}
            components={{
              body: {
                row: Row,
              },
            }}
            rowKey="key"
            columns={columns}
            dataSource={dataSource}
          />
        </SortableContext>
      </DndContext>
      <TextButton text="+" onClick={add} />
    </div>
  );
}
