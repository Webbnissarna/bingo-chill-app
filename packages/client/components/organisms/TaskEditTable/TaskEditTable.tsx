import {
  DeleteOutlined,
  MenuOutlined,
  PlusCircleOutlined,
} from "@ant-design/icons";
import type { DragEndEvent } from "@dnd-kit/core";
import { DndContext } from "@dnd-kit/core";
import { restrictToVerticalAxis } from "@dnd-kit/modifiers";
import {
  arrayMove,
  SortableContext,
  useSortable,
  verticalListSortingStrategy,
} from "@dnd-kit/sortable";
import { CSS } from "@dnd-kit/utilities";

import { Table } from "antd";
import type { ColumnsType } from "antd/es/table";
import React, { useEffect, useMemo, useState } from "react";
import type { Task } from "@webbnissarna/bingo-chill-common/src/game/types";
import { TagsInput, Button, TextInputField } from "@/components/atoms";
import ImagePicker from "@/components/molecules/ImagePicker";
import { nanoid } from "nanoid";
import type { Tag } from "@/components/atoms/TagsInput";
import { patch } from "@webbnissarna/bingo-chill-common/src/utils/functional";

export interface TaskWithKey extends Task {
  key: string;
}

interface RowProps extends React.HTMLAttributes<HTMLTableRowElement> {
  "data-row-key": string;
}

function Row({ children, ...props }: RowProps) {
  const {
    attributes,
    listeners,
    setNodeRef,
    setActivatorNodeRef,
    transform,
    transition,
    isDragging,
  } = useSortable({
    id: props["data-row-key"],
  });

  const style: React.CSSProperties = {
    ...props.style,
    transform: CSS.Transform.toString(transform && { ...transform, scaleY: 1 }),
    transition,
    ...(isDragging ? { position: "relative", zIndex: 9999 } : {}),
  };

  return (
    <tr {...props} ref={setNodeRef} style={style} {...attributes}>
      {React.Children.map(children, (child) => {
        if ((child as React.ReactElement).key === "sort") {
          return React.cloneElement(child as React.ReactElement, {
            children: (
              <MenuOutlined
                ref={setActivatorNodeRef}
                style={{ touchAction: "none", cursor: "move" }}
                {...listeners}
              />
            ),
          });
        }
        return child;
      })}
    </tr>
  );
}

function TaskToDateType(task: Task | TaskWithKey): TaskWithKey {
  return { ...task, key: (task as TaskWithKey).key ?? nanoid() };
}

export interface TaskEditTableProps {
  tasks: Task[];
  onChanged: (newTasks: TaskWithKey[]) => void;
}

export function stripKey(task: TaskWithKey | Task): Task {
  return { ...task, key: undefined } as unknown as Task;
}

export default function TaskEditTable({
  tasks,
  onChanged,
}: TaskEditTableProps): JSX.Element {
  const [dataSource, setDataSource] = useState<TaskWithKey[]>([]);

  useEffect(() => {
    setDataSource(tasks.map(TaskToDateType));
  }, [tasks, setDataSource]);

  const onDragEnd = ({ active, over }: DragEndEvent) => {
    if (active.id !== over?.id) {
      const activeIndex = dataSource.findIndex((i) => i.key === active.id);
      const overIndex = dataSource.findIndex((i) => i.key === over?.id);
      const changed = arrayMove(dataSource, activeIndex, overIndex);
      setDataSource(changed);
      onChanged(changed);
    }
  };

  const add = () => {
    onChanged([...dataSource, { key: nanoid(), name: "", icon: "", tags: [] }]);
  };

  const del = (key: string) => {
    onChanged(dataSource.filter((v) => v.key !== key));
  };

  const edit = (key: string, update: Partial<TaskWithKey>) => {
    onChanged(
      dataSource.map((item) => (item.key === key ? patch(item, update) : item)),
    );
  };

  const allTags = useMemo<Tag[]>(
    () =>
      [...new Set(tasks.map((t) => t.tags).flat())].map((value) => ({
        value,
        label: value,
      })),
    [tasks],
  );

  const columns: ColumnsType<TaskWithKey> = [
    {
      width: 50,
      key: "sort",
      align: "center",
    },
    {
      title: "Icon",
      dataIndex: "icon",
      width: 100,
      render: (_, { key, icon }) => (
        <ImagePicker
          value={icon}
          onChange={(newIcon) => edit(key, { icon: newIcon ?? undefined })}
        />
      ),
    },
    {
      title: "Name",
      dataIndex: "name",
      render: (_, { key, name }) => (
        <TextInputField
          value={name}
          onChange={(newName) => edit(key, { name: newName })}
        />
      ),
    },
    {
      title: "Tags",
      dataIndex: "tags",
      render: (_, { key, tags }) => (
        <TagsInput
          values={tags}
          onChange={(newTags) => edit(key, { tags: newTags })}
          options={allTags}
          placeholder="tags"
        />
      ),
    },
    {
      width: 70,
      render: (_, { key }) => (
        <Button variant="dangerous" onClick={() => del(key)}>
          <DeleteOutlined />
        </Button>
      ),
    },
  ];

  return (
    <div className="flex flex-col gap-2">
      <DndContext modifiers={[restrictToVerticalAxis]} onDragEnd={onDragEnd}>
        <SortableContext
          items={dataSource.map((i) => i.key)}
          strategy={verticalListSortingStrategy}
        >
          <Table
            pagination={false}
            components={{
              body: {
                row: Row,
              },
            }}
            rowKey="key"
            columns={columns}
            dataSource={dataSource}
          />
        </SortableContext>
      </DndContext>
      <Button onClick={add}>
        <PlusCircleOutlined />
      </Button>
    </div>
  );
}
