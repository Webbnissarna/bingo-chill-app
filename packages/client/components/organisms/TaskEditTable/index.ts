export { default } from "./TaskEditTable";
export * from "./TaskEditTable";
