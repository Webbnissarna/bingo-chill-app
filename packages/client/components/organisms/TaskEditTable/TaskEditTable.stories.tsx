import type { StoryDefault } from "@ladle/react";
import TaskEditTable from ".";
import { useState } from "react";
import type { Task } from "@webbnissarna/bingo-chill-common/src/game/types";

export function Default() {
  const [tasks, setTasks] = useState<Task[]>([
    { name: "Hello World", icon: "", tags: ["A", "B", "C"] },
    { name: "Hello World 2", icon: "", tags: ["A", "B", "D", "E"] },
  ]);
  return (
    <div className="bg-polarNight-1 p-5 flex flex-col gap-3">
      <TaskEditTable tasks={tasks} onChanged={setTasks} />
      <div className="bg-polarNight-0 p-2 overflow-x-scroll">
        <pre className="font-monospace text-snowStorm-0">
          {JSON.stringify(tasks, null, 2)}
        </pre>
      </div>
    </div>
  );
}

export default {
  title: "Organisms / Task Edit Table",
} satisfies StoryDefault;
