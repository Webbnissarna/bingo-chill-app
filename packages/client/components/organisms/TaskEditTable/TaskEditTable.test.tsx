import { render } from "../../../.jest/render";
import TaskEditTable from ".";

describe("TaskEditTable", () => {
  it("renders", () => {
    const { asFragment } = render(
      <TaskEditTable tasks={[]} onChanged={() => undefined} />,
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
