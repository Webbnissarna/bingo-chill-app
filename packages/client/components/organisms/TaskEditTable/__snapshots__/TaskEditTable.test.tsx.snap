// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskEditTable renders 1`] = `
<DocumentFragment>
  <div
    class="flex flex-col gap-2"
  >
    <div
      class="ant-table-wrapper css-dev-only-do-not-override-18iikkb"
    >
      <div
        class="ant-spin-nested-loading css-dev-only-do-not-override-18iikkb"
      >
        <div
          class="ant-spin-container"
        >
          <div
            class="ant-table ant-table-empty"
          >
            <div
              class="ant-table-container"
            >
              <div
                class="ant-table-content"
              >
                <table
                  style="table-layout: auto;"
                >
                  <colgroup>
                    <col
                      style="width: 50px;"
                    />
                    <col
                      style="width: 100px;"
                    />
                    <col />
                    <col />
                    <col
                      style="width: 70px;"
                    />
                  </colgroup>
                  <thead
                    class="ant-table-thead"
                  >
                    <tr>
                      <td
                        class="ant-table-cell"
                        style="text-align: center;"
                      />
                      <th
                        class="ant-table-cell"
                        scope="col"
                      >
                        Icon
                      </th>
                      <th
                        class="ant-table-cell"
                        scope="col"
                      >
                        Name
                      </th>
                      <th
                        class="ant-table-cell"
                        scope="col"
                      >
                        Tags
                      </th>
                      <td
                        class="ant-table-cell"
                      />
                    </tr>
                  </thead>
                  <tbody
                    class="ant-table-tbody"
                  >
                    <tr
                      aria-describedby="DndDescribedBy-0"
                      aria-disabled="false"
                      aria-pressed="true"
                      aria-roledescription="sortable"
                      class="ant-table-placeholder"
                      role="button"
                      style="position: relative; z-index: 9999; transition: transform 0ms linear;"
                      tabindex="0"
                    >
                      <td
                        class="ant-table-cell"
                        colspan="5"
                      >
                        <div
                          class="css-dev-only-do-not-override-18iikkb ant-empty ant-empty-normal"
                        >
                          <div
                            class="ant-empty-image"
                          >
                            <svg
                              height="41"
                              viewBox="0 0 64 41"
                              width="64"
                              xmlns="http://www.w3.org/2000/svg"
                            >
                              <g
                                fill="none"
                                fill-rule="evenodd"
                                transform="translate(0 1)"
                              >
                                <ellipse
                                  cx="32"
                                  cy="33"
                                  fill="#f5f5f5"
                                  rx="32"
                                  ry="7"
                                />
                                <g
                                  fill-rule="nonzero"
                                  stroke="#d9d9d9"
                                >
                                  <path
                                    d="M55 12.76L44.854 1.258C44.367.474 43.656 0 42.907 0H21.093c-.749 0-1.46.474-1.947 1.257L9 12.761V22h46v-9.24z"
                                  />
                                  <path
                                    d="M41.613 15.931c0-1.605.994-2.93 2.227-2.931H55v18.137C55 33.26 53.68 35 52.05 35h-40.1C10.32 35 9 33.259 9 31.137V13h11.16c1.233 0 2.227 1.323 2.227 2.928v.022c0 1.605 1.005 2.901 2.237 2.901h14.752c1.232 0 2.237-1.308 2.237-2.913v-.007z"
                                    fill="#fafafa"
                                  />
                                </g>
                              </g>
                            </svg>
                          </div>
                          <div
                            class="ant-empty-description"
                          >
                            No data
                          </div>
                        </div>
                      </td>
                    </tr>
                  </tbody>
                </table>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      id="DndDescribedBy-0"
      style="display: none;"
    >
      
    To pick up a draggable item, press the space bar.
    While dragging, use the arrow keys to move the item.
    Press space again to drop the item in its new position, or press escape to cancel.
  
    </div>
    <div
      aria-atomic="true"
      aria-live="assertive"
      id="DndLiveRegion-0"
      role="status"
      style="position: fixed; width: 1px; height: 1px; margin: -1px; border: 0px; padding: 0px; overflow: hidden; clip-path: inset(100%); white-space: nowrap;"
    />
    <button
      class="font-text font-base text-snowStorm-2 px-3 py-1 rounded-lg transition flex gap-1 items-center justify-center bg-polarNight-3 hover:bg-frost-3 active:bg-frost-2"
    >
      <span
        aria-label="plus-circle"
        class="anticon anticon-plus-circle"
        role="img"
      >
        <svg
          aria-hidden="true"
          data-icon="plus-circle"
          fill="currentColor"
          focusable="false"
          height="1em"
          viewBox="64 64 896 896"
          width="1em"
        >
          <path
            d="M696 480H544V328c0-4.4-3.6-8-8-8h-48c-4.4 0-8 3.6-8 8v152H328c-4.4 0-8 3.6-8 8v48c0 4.4 3.6 8 8 8h152v152c0 4.4 3.6 8 8 8h48c4.4 0 8-3.6 8-8V544h152c4.4 0 8-3.6 8-8v-48c0-4.4-3.6-8-8-8z"
          />
          <path
            d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm0 820c-205.4 0-372-166.6-372-372s166.6-372 372-372 372 166.6 372 372-166.6 372-372 372z"
          />
        </svg>
      </span>
    </button>
  </div>
</DocumentFragment>
`;
